Optimize RLE by 40%

Kubuxu opened the pull request:
It is a bit slower for very small RLEs (10 bits or less) but 41% faster for decode+encode
loop with real, big RLE from chain.

Full bench run:
```
Count/0/basic-8              192ns ±10%   218ns ±15%  +13.22%  (p=0.009 n=10+10)
Count/1/basic-8              187ns ±13%   207ns ± 6%  +10.64%  (p=0.001 n=10+9)
Count/1/modified-8           868ns ±11%   962ns ± 6%  +10.77%  (p=0.002 n=10+7)
Count/10/basic-8             797ns ± 6%   823ns ± 6%     ~     (p=0.123 n=10+10)
Count/10/modified-8         1.45µs ± 3%  1.27µs ± 8%  -12.70%  (p=0.000 n=10+10)
Count/1000/basic-8          90.0µs ± 1%  74.5µs ± 7%  -17.29%  (p=0.000 n=9+9)
Count/1000/modified-8       94.5µs ± 0%  82.0µs ± 0%  -13.17%  (p=0.000 n=8+8)
Count/1000000/basic-8        141ms ± 1%   126ms ± 3%  -10.67%  (p=0.000 n=8+9)
Count/1000000/modified-8     146ms ± 1%   128ms ± 3%  -12.16%  (p=0.000 n=9+10)
IsEmpty/0/basic-8            217ns ± 4%   217ns ± 8%     ~     (p=0.591 n=10+10)
IsEmpty/1/basic-8            218ns ± 5%   247ns ±16%  +13.14%  (p=0.000 n=10+10)
IsEmpty/1/modified-8        1.00µs ± 2%  1.04µs ± 6%   +4.03%  (p=0.002 n=7+10)
IsEmpty/10/basic-8           804ns ± 2%   818ns ± 3%     ~     (p=0.161 n=9+9)
IsEmpty/10/modified-8       1.40µs ± 9%  1.45µs ±10%     ~     (p=0.197 n=10+10)
IsEmpty/1000/basic-8        70.9µs ± 1%  70.6µs ± 2%     ~     (p=0.093 n=8+9)
IsEmpty/1000/modified-8     71.7µs ± 1%  61.4µs ± 1%  -14.39%  (p=0.000 n=9+9)
IsEmpty/1000000/basic-8      124ms ± 2%   107ms ± 1%  -13.35%  (p=0.000 n=9+9)
IsEmpty/1000000/modified-8   124ms ± 2%   107ms ± 0%  -13.55%  (p=0.000 n=10+9)
BigDecodeEncode-8           1.37ms ±11%  0.80ms ± 1%  -41.87%  (p=0.000 n=10+9)
```

Kubuxu committed:
Add test vector from chain

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu committed:
Optimize RLE decoding by decode table

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu committed:
Eliminate NormIter from EncodeRuns

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Stebalien commented on rle/bitvec.go:
```diff
@@ -49,6 +49,14 @@ func (bv *rbitvec) GetByte() byte {
+}
+
+func (bv *rbitvec) Drop(num byte) {
+	bv.Get(num)
```
Does specializing this make it any faster?

Kubuxu committed:
Remove missleading comment

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -4,6 +4,55 @@ import (
+	}
+
+	for i := 1; i <= 6; i++ {
+		for j := 0; j < 1<<6>>i; j++ {
```
Parentheses please.

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -4,6 +4,55 @@ import (
+	varint bool
+}
+
+var decodeTable = [1 << 6]decodeInfo{}
```
Big comment explaining what this is and how it works.

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -21,33 +70,45 @@ func DecodeRLE(buf []byte) (RunIterator, error) {
+	it.lastVal = ret.Val
+
+	it.i--
+	if it.i == 255 { // if i was 0 before subtraction
```
Does this really matter for performance?;

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -4,6 +4,55 @@ import (
+		}
+	}
+}
+
```
Can we have an exhaustive test for all 64 of these values?

Stebalien reviewed (commented):
I think this is correct, but it's hard to tell just by reading it. It would be nice to have an exhaustive test to be sure.

Kubuxu committed:
Use Zipf distributed RLEs for benchmarks

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu committed:
Make it 5.6% faster

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu committed:
Add test for all values in decode table

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu commented:
@Stebalien is the test testing all values in the decode table enough? or do you want the reader to actually decode these runs?

Stebalien commented:
I'm more interested in actually decoding them. I trust we'll generate the table correctly, I just want to make sure correctly is the correct correctly.

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -4,6 +4,59 @@ import (
+	// runs of 1s 0bxxxxxx1, 0bxxxx11 ...
+	for i := 1; i <= 6; i++ {
+		for j := 0; j < (1<<6)>>i; j++ {
+			idx := bitMasks[i] | byte(j<<i)
+			decodeTable[idx] = decodeInfo{
+				length: 1,
+				i:      byte(i - 1),
+				n:      byte(i),
```
I believe this produces a table that works, but it's not the optimal table. For example:

```
i=1
j=1
idx=0b000011
```

We'll end up treating the run as length 1 instead of length 2, right?

Stebalien replied:
What if we iterated over all `i in 1<<6` and just decoded these numbers? That seems like a more reliable approach.

Kubuxu replied:
it will be two repeats (`i` is 0 based) of run length of 1.
The higher number of repeats are set latter so they override the lower number of repeats with more specific decode info.

Iterating over the whole range and decoding makes sense, but I'm not sure what is more complex.

Stebalien replied:
Ah, got it. I did not expect this to overwrite.

Kubuxu committed:
Redefine decode table by iterating over all entries

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -1,9 +1,57 @@
+// 0bxxxx00 - varint run, the decode value not defined in LUT
+var decodeTable = [1 << 6]decodeInfo{}
+
+func buildDecodeTable() {
```
This is very easy for me to read now.

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -1,9 +1,57 @@
+	length byte // length of the run
+	i      byte // i+1 is number of repeats of above run lengths
+	n      byte // number of bits to read
+	varint bool // varint signifies that futher bits need to be processed as a varint
```
So, _technically_ we don't need this last flag as we can say "length == 0" means "varint". But that's probably too tricky.

Kubuxu replied:
yeah, but to gain anything we would have to get it all to fit in 16 bits, so not worth it right now.

Stebalien replied:
Well, my thought was this plus nibbles. But yeah, not worth it.

Stebalien commented on rle/rleplus_reader.go:
```diff
@@ -1,9 +1,57 @@
 )
 
+type decodeInfo struct {
+	length byte // length of the run
+	i      byte // i+1 is number of repeats of above run lengths
+	n      byte // number of bits to read
```
Followup: nibbles?

Kubuxu replied:
Sorry, I don't follow.

Kubuxu replied:
If you mean packing it in 4 bit fragments, possible (I have a copy locally) but it is only 0.5% faster so not really worth the complexity.

Stebalien replied:
Yes. Oh well... I guess we'll just have to live with 40%. :man_shrugging:

Kubuxu committed:
Add test for all 6 bit values

```
=== RUN   TestAll8bit
000000: 34
000001: 6
000010: 0
000011: 12
000100: 33
000101: 1
000110: 0
000111: 24
001000: 33
001001: 6
001010: 6
001011: 3
001100: 33
001101: 1
001110: 7
001111: 48
010000: 33
010001: 6
010010: 7
010011: 12
010100: 33
010101: 1
010110: 7
010111: 7
011000: 33
011001: 6
011010: 7
011011: 3
011100: 33
011101: 2
011110: 7
011111: 62
100000: 33
100001: 6
100010: 7
100011: 12
100100: 33
100101: 2
100110: 7
100111: 24
101000: 33
101001: 6
101010: 7
101011: 4
101100: 33
101101: 2
101110: 7
101111: 15
110000: 33
110001: 6
110010: 7
110011: 12
110100: 33
110101: 2
110110: 7
110111: 7
111000: 33
111001: 6
111010: 7
111011: 4
111100: 32
111101: 2
111110: 7
111111: 66
```

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

Kubuxu committed:
lint: delete bitMasks

Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>
